pipeline/beam_tables.py

```diff
@@ -448,15 +386,151 @@ def _flatten_measurement(
+        'ip': scan['resolver'],
+        'date': date,
+        'error': scan.get('error', None),
+        'blocked': not scan['passed'] if 'passed' in scan else None,
+        'success': 'error' not in scan,
+        'received': None,
+        'measurement_id': random_measurement_id,
+    }
+    received_ips = scan.get('answers')
+    yield from self._process_received_ips(row, received_ips)
+
+  def _process_satellite_v2(self, scan: Any,
+                            random_measurement_id: str) -> Iterator[Row]:
+    """Process a line of Satellite data.
+
+    Args:
+      scan: a loaded json object containing the parsed content of the line
+      random_measurement_id: a hex id identifying this individual measurement
+
+    Yields:
+      Rows
+    """
+    row = {
+        'domain': scan['test_url'],
+        'ip': scan['vp'],
+        'country': scan['location']['country_code'],
+        'date': scan['start_time'][:10],
+        'start_time': scan['start_time'],
+        'end_time': scan['end_time'],
+        'error': scan.get('error', None),
+        'blocked': scan['anomaly'],
+        'success': not scan['connect_error'],
+        'received': None,
+        'measurement_id': random_measurement_id
+    }
+    received_ips = scan.get('response')
+    yield from self._process_received_ips(row, received_ips)
+
+  def _process_received_ips(  # pylint: disable=no-self-use
+      self, row: Row, received_ips: Optional[Dict[str, str]]) -> Iterator[Row]:
+    """Add received_ip metadata to a Satellite row
+
+    Args:
+      row: existing row of satellite data
+      received_ips: data on the response from the resolver (error or ips)
+
+    Yields:
+      Rows
+    """
+    if not received_ips:
+      yield row
+      return
+
+    # separate into one answer ip per row for tagging
+    if 'rcode' in received_ips:
+      row['rcode'] = received_ips.pop('rcode', None)
+    for ip in received_ips:
+      row['received'] = {'ip': ip}
+      if row['blocked']:
+        # Track domains per IP for interference
+        INTERFERENCE_IPDOMAIN[ip].add(row['domain'])
+      if isinstance(received_ips, dict):
+        row['received']['matches_control'] = ' '.join(  # pylint: disable=unsupported-assignment-operation
+            [tag for tag in received_ips[ip] if tag in SATELLITE_TAGS])
+      yield row
+
+  def _parse_received_data(self, received: Union[str, Dict[str, Any]],
+                           anomaly: bool) -> Row:
+    """Parse a received field into a section of a row to write to bigquery.
+
+    Args:
+      received: a dict parsed from json data, or a str
+      anomaly: whether data may indicate blocking
+
+    Returns:
+      a dict containing the 'received_' keys/values in SCAN_BIGQUERY_SCHEMA
+    """
+    if isinstance(received, str):
+      return {'received_status': received}
+
+    row = {
+        'received_status':
+            received['status_line'],
+        'received_body':
+            received['body'],
```

Comment:
I'm surprised this doesn't crash with a KeyError.

Reply:
Some of the fields are just very consistent and exist in 100% of the data.